mansrc/bam2wig.1
'\" t
.\"     Title: bam2wig
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "BAM2WIG" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
bam2wig \- Convert bam files to wiggle files. Can be used to generate exonpart hints for AUGUSTUS.
.SH "SYNOPSIS"
.sp
\fBbam2wig\fP [\-r region] [\-t trackname] <in.bam>
.SH "OPTIONS"
.sp
File needs to be sorted by Reference ID (i.e. target name)
Use "samtools sort <in.bam>" to such effect.
Use "samtools stats <in.bam> | grep \(aqis sorted\(aq " to check if the file is sorted.
.SS "Parameters:"
.sp
\fB\-r\fP
.RS 4
Allows to specify a target region, e.g. \(aqchr3L:10\-250\(aq
This option can only be used if an index file exists, see: samtools index
.RE
.sp
\fB\-t\fP
.RS 4
string provided as track name
.RE
.SH "AUTHORS"
.sp
AUGUSTUS was written by M. Stanke, O. Keller, S. König, L. Gerischer and L. Romoth.
.SH "ADDITIONAL DOCUMENTATION"
.sp
An exhaustive documentation can be found in the file /usr/share/doc/augustus/README.md.gz.

HISTORY.TXT
     - remove bam2wig
     - compile with CGP and ZIPINPUT support by default
     - compile with SQLite and MySQL support for CGP (switch of by adding SQLite = false and MySQL = false to common.mk)
List of changes from version 3.3.2 to 3.3.3 (until Sep 13th, 2019)
     - docker container
     - new script tfix_in_frame_stop_codon_genes.py that replaces genes where spliced in-frame stop codons were predicted
     - new scripts: compare_masking.pl, merge_masking.pl
     - new program bamToWig.py as alternarive to bam2wig C++ binary
     - fix warnings on new GCC compiler (8.3)
     - introduced unit tests of C++ code (make test)
     - bugfixes (lex.cc, autoAug.pl)
     - DIAMOND as alternative to BLAST in aa2nonred.pl
List of changes from version 3.3.1 to 3.3.2 (until Oct 5th, 2018)
     - bugfixes in comparative augustus, utrrnaseq
     - new species Chiloscyllium punctatum (bamboo shark), Scyliorhinus torazame (cat shark), Rhincodon typus (whale shark)
     - updated comparative augustus (CGP) tutorial 
List of changes from version 3.3 to 3.3.1 (until May 8th, 2018)
     - new species pisaster (Pisaster ochraceus, ochre starfish)
     - bugfixes of sampling error in intron model and in check of
       frame-compatibility of CDSpart hints without frame
     - new auxiliary program utrrnaseq to find training UTRs
     - new script setStopCodonFreqs.pl
List of changes from version 3.2.3 to 3.3 (until July 11th, 2017)
     - new program ESPOCA to estimate selective pressure on codon alignments
     - gene finding on ancestral genomes is enabled
     - new default parameters for comparative gene prediction (CGP)
     - clade parameters training for CGP
     - compatibility to Ian Fiddes' Comparative Annotation Toolkit (CAT)
     - new scripts eval_dualdecomp.pl,
     - more tolerant tree parsing
     - bugfixes in augustus, joingenes, load2sqlitedb, transMap2hints.pl, splitMfasta.pl, intron2exex.pl,
       aln2wig
     - new functionality in homGeneMapping, joingenes
     - new species ciona, mnemiopsis_leidyi, nematostella_vectensis, strongylocentrotus_purpuratus
List of changes from version 3.2.2 to 3.2.3 (until December 6th, 2016)
     - bugfixes in joingenes, load2sqlitedb, transMap2hints.pl
     - added script gtf2bed.pl
     - CGP parameters for clade of 8 primates (cgp/log_reg_parameters_primates.cfg)
     - added option --clean to load2sqlitedb
     - now filter out truncated coding genes with CDS length <= 3
     - Sascha Steinbiss: remove dependency on PATH_MAX
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic license only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently

auxprogs/bam2wig/bam2wig.c
/* 
   Creates a Wiggle file with coverage information coming from a BAM file 	
 																			
   NOTE: 
   Depending on the version of the compiler,  the call to "-lcurses" might have 
   to be replaced to "-lncurses"
 					
   Tonatiuh Pena-Centeno														
   Created: 12-June-2012 													
   Last modified:   6-November-2012												
*/


#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#include "bgzf.h"
#include "sam.h"

// Auxiliary data structure
typedef struct {
	htsFile *hts_fp; // the file handler
	BGZF *bgzf_fp;   // the file handler
	hts_itr_t *iter; // NULL if a region not specified
	int min_mapQ;    // mapQ (for filtering purposes but not used in this app)
} aux_t;


// Reads a BAM alignment from a BAM file.
static int read_bam(void *data, bam1_t *b) 
{// If necessary, include UNMAPPED filtering in here to avoid pileup (use min_mapQ)

	// data is a pointer to the auxiliary structure
	aux_t *aux = (aux_t*)data; 

	// Compute coverage according to the specified region, if one has been provided (i.e. bam_iter_read)
	// or compute coverage of the complete alignment otherwise (bam_read1)
	int ret = aux->iter? hts_itr_next(aux->bgzf_fp, aux->iter, b, aux->hts_fp) : bam_read1(aux->bgzf_fp, b);

	return ret;
}

void usage()
{
  printf("\n");
  fprintf(stderr, "Usage: bam2wig [-r region] [-t trackname] <in.bam> \n");
  printf("-----------------------------------------------------------------\n");
  printf(" -r   Allows user to specify a target region, e.g. 'chr3L:10-250'\n");
  printf("      This option can only be used if an index file exists\n");
  printf("      See: samtools index \n");	
  printf(" -t   A string might be provided as track name\n");
  printf("\n");	
  printf("NOTE:");
  printf("File needs to be sorted by Reference ID (i.e. target name)\n");
  printf("Use 'samtools sort <in.bam>' to such effect.\n"); 
  printf("\n");
  exit(1);
}

int main(int argc, char *argv[])
{

	char *filename=NULL;
	char *trackname=NULL;

	int n, tid, *n_plp;
	hts_pos_t beg, end, pos;
	const bam_pileup1_t **plp;
	char *reg = 0; // specified region
	aux_t **data;
	bam_mplp_t mplp;

	// Parsing the command line
	while ((n = getopt(argc, argv, "r:t:")) >= 0) 
	  {
		switch (n) 
			{
			  case 'r': reg = strdup(optarg); break;   // parsing a region requires a BAM header
			  case 't': trackname = strdup(optarg); break;    // mapping quality threshold
			  default :
				usage();
			}
	  }

	if (optind == argc || (argc-optind) != 1)
	  {
		usage();
	  }


	// Initializing auxiliary data structures
	data = calloc(1, sizeof(void*)); // data[0] is array for just one BAM file
	// set the default region to the maximum value of hts_pos_t
	beg = 0; end = HTS_POS_MAX; tid = -1;

	// Opening BAM file
	char *oldTargetName = "", *newTargetName;
	filename = argv[optind];
	data[0] = calloc(1, sizeof(aux_t));
	data[0]->hts_fp = hts_open(filename, "r");              // file handler of BAM
	if (data[0]->hts_fp == NULL) {
		fprintf(stderr, "Failed to open file \"%s\" : No such file or directory or not a bam file.\n", filename);
		exit(1);
	}
	if (data[0]->hts_fp->format.format != bam) {
		fprintf(stderr, "File \"%s\" is not in bam file format.\n", filename);
		exit(1);
	}
	data[0]->bgzf_fp = data[0]->hts_fp->fp.bgzf;            // file handler of BAM
	data[0]->min_mapQ = 0;                    		// mapQ is not used by this app
	// Reading BAM header
	sam_hdr_t *htmp = bam_hdr_read(data[0]->bgzf_fp);

	// parsing region
	if (reg) 
		{ 
			hts_parse_region(reg, &tid, &beg, &end, (hts_name2id_f)bam_name2id, htmp, 0);
		}

	if (tid >= 0) 
	  { // if a region is specified and parsed successfully
		hts_idx_t *idx = hts_idx_load(argv[optind], HTS_FMT_BAI);  // load the index

		if (idx == NULL)
		  {
		    fprintf(stderr, "Missing indexed BAM file!\n");
		    fprintf(stderr, "See: samtools index\n");
		    exit(1);
		  }

		data[0]->iter = sam_itr_queryi(idx, tid, beg, end); // set the iterator
		hts_idx_destroy(idx); // the index is not needed any more; phase out of the memory
	  }


	// The set of multi-pileup functions are used to obtain the coverage information 
	mplp = bam_mplp_init(1, read_bam, (void**)data); // initialization
	n_plp = calloc(1, sizeof(int)); // n_plp[0] contains the number of covering reads 
	plp = calloc(1, sizeof(void*)); // plp[0] points to the array of covering reads (internal in mplp)

	// Print default trackname or use the one specified by the user
	printf("track name=%s type=wiggle_0\n", trackname==NULL? filename : trackname);


	while (bam_mplp64_auto(mplp, &tid, &pos, n_plp, plp) > 0)
	  { // come to the next covered position

		// If requested region is of range, skip
		if (pos < beg || pos >= end) 
		 	continue;

		// Print the reference name as the header of each track
		newTargetName = htmp->target_name[tid];
		if (strcmp(oldTargetName, newTargetName))
		  { 
			printf("variableStep chrom=%s\n", newTargetName); 
		  }

		// Verifying whether the array of covering reads corresponds to "del" or "refskip"
		int j, m = 0;
		for (j = 0; j < n_plp[0]; ++j)
		  {
			const bam_pileup1_t *p = plp[0] + j; 	// DON'T modify plp[][] unless you really know
			if (p->is_del || p->is_refskip) 
			  {// having dels or refskips at tid:pos
				++m;
			  } 	
		  }

		// Estimates the depth, discounting "m" from the pile if: dels/refskips were found
		int coverage = n_plp[0] - m;

		// Prints position and coverage
		if (coverage > 0) 
		  {
			printf("%ld %d\n", pos+1, coverage);
		  }

		// Update reference name
		oldTargetName = newTargetName;

	  } // end while


	// Freeing used memory and closing handles
	free(n_plp); 
	free(plp);
	bam_mplp_destroy(mplp);
	sam_hdr_destroy(htmp);
	hts_close(data[0]->hts_fp);

	// Iterator is used only when a region was provided
	if (data[0]->iter) 
	  { 
		hts_itr_destroy(data[0]->iter); 
	  }
	free(data[0]); 
	free(data); 
	free(reg);

	return 0;
}
